python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,753 @@
+        numpy.float64: DataType.FP64,
+        numpy.bytes_: DataType.BYTES,
+        numpy.str_: DataType.BYTES,
+        numpy.object_: DataType.BYTES,
+    },
+)
+
+TRITON_TO_NUMPY_DTYPE: dict[DataType, type] = _CustomKeyErrorDict(
+    "Triton data type",
+    "Numpy data type",
+    {
+        **{value: key for key, value in NUMPY_TO_TRITON_DTYPE.items()},
+        **{DataType.BYTES: numpy.object_},
+    },
+)
+
+
+@dataclass
+class MemoryBuffer:
+    """Memory allocated for a Tensor"""
+
+    data_ptr: int
+    memory_type: MemoryType
+    memory_type_id: int
+    size: int
+    owner: Any
+
+    @staticmethod
+    def from_dlpack(
+        owner: Any, dlpack_object: Optional[DLPackObject] = None
+    ) -> MemoryBuffer:
+        if dlpack_object is None:
+            if not hasattr(owner, "__dlpack__"):
+                raise InvalidArgumentError("Object does not support DLpack protocol")
+
+            dlpack_object = DLPackObject(owner)
+
+        if not dlpack_object.contiguous:
+            raise InvalidArgumentError("Only contiguous memory is supported")
+
+        return MemoryBuffer(
+            int(dlpack_object.data_ptr),
+            dlpack_object.memory_type,
+            dlpack_object.memory_type_id,
+            dlpack_object.byte_size,
+            owner,
+        )
+
+    def _create_TRITONSERVER_BufferAttributes(self) -> BufferAttributes:
+        buffer_attributes = BufferAttributes()
+        buffer_attributes.memory_type = self.memory_type
+        buffer_attributes.memory_type_id = self.memory_type_id
+        buffer_attributes.byte_size = self.size
+        #        buffer_attributes.cuda_ipc_handle = None
+        return buffer_attributes
+
+
+class MemoryAllocator(ABC):
+    @abstractmethod
+    def allocate(
+        self, size: int, memory_type: MemoryType, memory_type_id: int, tensor_name: str
+    ) -> MemoryBuffer:
+        """Allocate memory buffer for tensor.
+
+        Parameters
+        ----------
+        size : int
+            number of bytes requested
+        memory_type : MemoryType
+                type of memory requested (CPU, GPU, etc.)
+        memory_type_id : int
+            memory type id requested (typically device id)
+        tensor_name : str
+            name of tensor
+
+        Returns
+        -------
+        MemoryBuffer
+            memory buffer with requested size
+
+        Examples
+        --------
+        memory_buffer = allocator.allocate(100,MemoryType.CPU,0,"output")
+        """
+
+        pass
+
+
+class NumpyAllocator(MemoryAllocator):
+    def __init__(self):
+        pass
+
+    def allocate(
+        self, size: int, memory_type: MemoryType, memory_type_id: int, tensor_name: str
+    ) -> MemoryBuffer:
+        ndarray = numpy.empty(size, numpy.byte)
+        return MemoryBuffer.from_dlpack(ndarray)
+
+
+default_memory_allocators: dict[MemoryType, MemoryAllocator] = dict(
+    {
+        MemoryType.CPU: NumpyAllocator(),
+    }
+)
+
+if cupy is not None:
```

Comment:
I am still not fully onboard with having cupy as one of the dependencies for Python APIs. I think using `cuda-python` probably makes more sense since we're just trying to allocate memory.